File: packages/@internal/descope/src/store.ts
import { Store, type Collection } from "@internal/core";
import type { DescopeUser, DescopeOAuthClient } from "./entities.js";

export interface DescopeStore {
  users: Collection<DescopeUser>;
  oauthClients: Collection<DescopeOAuthClient>;
}

export function getDescopeStore(store: Store): DescopeStore {
  return {
    users: store.collection<DescopeUser>("descope.users", ["uid", "email"]),
    oauthClients: store.collection<DescopeOAuthClient>("descope.oauth_clients", ["client_id"]),
  };
}

File: internal/services/microsoft/store.go
package microsoft

import corestore "github.com/vercel-labs/emulate/internal/core/store"

type Store struct {
	Users         *corestore.Collection
	OAuthClients  *corestore.Collection
	OAuthCodes    *corestore.Collection
	RefreshTokens *corestore.Collection
	AccessTokens  *corestore.Collection
}

func NewStore(runtimeStore *corestore.Store) Store {
	return Store{
		Users:         runtimeStore.MustCollection("microsoft.users", "oid", "email"),
		OAuthClients:  runtimeStore.MustCollection("microsoft.oauth_clients", "client_id"),
		OAuthCodes:    runtimeStore.MustCollection("microsoft.oauth_codes", "code", "client_id", "email"),
		RefreshTokens: runtimeStore.MustCollection("microsoft.refresh_tokens", "token", "client_id", "email"),
		AccessTokens:  runtimeStore.MustCollection("microsoft.access_tokens", "token", "login"),
	}
}
